planner/core/rule_partition_processor.go

```diff
@@ -224,13 +224,80 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
 	return tableDual, nil
 }
 
-func (s *partitionProcessor) pruneListPartition(tbl table.Table, partitionNames []model.CIStr) []int {
-	if len(partitionNames) > 0 {
-		pi := tbl.Meta().Partition
+func (s *partitionProcessor) findUsedListPartitions(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	conds []expression.Expression) ([]int, error) {
+	pi := tbl.Meta().Partition
+	partExpr, err := tbl.(partitionTable).PartitionExpr()
+	if err != nil {
+		return nil, err
+	}
+	pruneList := partExpr.ForListPruning
+	cols := make([]*expression.Column, 0, len(pruneList.ExprCols))
+	colLen := make([]int, 0, len(pruneList.ExprCols))
+	for _, c := range pruneList.ExprCols {
```

Comment:
how about adding a test case like #13491, that list prune happen in the right child of  union or join

Reply:
Great catch, done. already add a test for this.